### PR TITLE
Aggregate function groupConcat (#63671)

### PR DESCRIPTION
**Pull request by yariks5s**
REVERTED

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
REVERTED

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All NOT Required Checks
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: Integration Tests
- [ ]  Exclude: Stateless tests
- [ ]  Exclude: Stateful tests
- [ ]  Exclude: Performance tests
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with Aarch64
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
---
- [ ]  Do not test
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit 2d189eb1a3973ecd8cdde12507bb9bafc6878da4 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/65009/2d189eb1a3973ecd8cdde12507bb9bafc6878da4/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>A Sync</td><td>If it fails, ask a maintainer for help</td><td>⏳ pending</td></tr>
<tr><td>PR Check</td><td>Checks correctness of the PR's body</td><td>❌ failure</td></tr>
<tbody>
</table>

File: src/AggregateFunctions/AggregateFunctionGroupConcat.cpp [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+#include <DataTypes/DataTypesNumber.h>
+
+#include <Columns/ColumnNullable.h>
+#include <Columns/ColumnString.h>
+
+#include <Common/ArenaAllocator.h>
+#include <Common/assert_cast.h>
+#include <Core/ColumnWithTypeAndName.h>
+#include <Columns/IColumn.h>
+#include <DataTypes/IDataType.h>
```
Comment (jsc0218) [inline src/AggregateFunctions/AggregateFunctionGroupConcat.cpp:20]: probably can reorder the headers here

**Comment by jsc0218** [comment]
might update with the master once, the test failure should irrelevant.

**Review by jsc0218 — APPROVED** [review]

File: src/AggregateFunctions/AggregateFunctionGroupConcat.cpp [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+    }
+
+    void serialize(ConstAggregateDataPtr __restrict place, WriteBuffer & buf, std::optional<size_t> /* version */) const override
+    {
+        auto & cur_data = this->data(place);
+
+        writeVarUInt(cur_data.data_size, buf);
+        writeVarUInt(cur_data.allocated_size, buf);
+
+        buf.write(cur_data.data, cur_data.data_size);
```
Comment (CheSema) [inline src/AggregateFunctions/AggregateFunctionGroupConcat.cpp:167]: I think the memory is not allocated yet here.
This is an example of SigFault https://s3.amazonaws.com/clickhouse-test-reports/64478/517c1138113220e5a4318c07806759911896234d/stress_test__debug_.html

**Comment by clickgapai** [comment]
Hi @yariks5s @Avogar @jsc0218 — while reviewing this PR I found the following:

- **Bug (P2):** `groupConcat` skips delimiters when leading elements are empty strings due to `data_size != 0` check — https://github.com/ClickHouse/ClickHouse/issues/104507

Happy to discuss — close anything that's wrong or already addressed.

**Pull request merged**